VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/ImplementEqualsAndGetHashCode/ImplementEqualsAndGetHashCodeAnalyzer.cs

```diff
@@ -23,38 +21,44 @@ public class ImplementEqualsAndGetHashCodeAnalyzer : DiagnosticAnalyzer
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext context) =>
-                context.RegisterSyntaxNodeAction(AnalyzeSymbol, SyntaxKind.ClassDeclaration, SyntaxKind.StructDeclaration);
+            context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
 
-        private void AnalyzeSymbol(SyntaxNodeAnalysisContext context)
+        private void AnalyzeSymbol(SymbolAnalysisContext symbol)
         {
             IMethodSymbol objectEquals;
             IMethodSymbol objectGetHashCode;
-            GetEqualsAndGetHashCodeSymbols(context.SemanticModel, out objectEquals, out objectGetHashCode);
 
-            var declaration = (TypeDeclarationSyntax)context.Node;
+            var namedType = (INamedTypeSymbol)symbol.Symbol;
+            GetEqualsAndGetHashCodeSymbols(namedType, out objectEquals, out objectGetHashCode);
 
-            if (!TypeMembersContainOverridenEqualsOrGetHashCode(context.SemanticModel, declaration.Members, objectEquals, objectGetHashCode) &&
-                MembersContainNonStaticFieldOrProperty(declaration.Members))
+            if (objectEquals != null || objectGetHashCode != null) { return; }
+
+            if (MembersContainNonStaticFieldOrProperty(namedType.GetMembers()))
             {
-                context.ReportDiagnostic(Diagnostic.Create(Rule, declaration.Identifier.GetLocation(),
-                    declaration.IsKind(SyntaxKind.ClassDeclaration) ? "Class" : "Struct", declaration.Identifier));
+                // it is only reported at the main location--the other locations are (apparently) just passed as data
+                // todo--ask the Roslyn team about this
+                for (var i = 0; i < namedType.Locations.Count(); i++)
+                {
+                    symbol.ReportDiagnostic(Diagnostic.Create(Rule, namedType.Locations[i],
+                    namedType.Locations.RemoveAt(i),
```

Comment:
So there's a warning for every partial declaration? I'm not entirely sure if that's what we want. What do you think about just picking the first location?